### PR TITLE
Auto add file extension when save - issue #1399

### PR DESCRIPTION
**Pull request by TomaszPluta**
Added auto add the file extension in "Save As" file dialog based on selected filter. The method " QFileDialog::getSaveFileName" in linux envrionment  doesn't give this possibility, so here is added  "ShowSaveFileDialog" class. 

This class performs what was needed in issue #1399:
_- If a graphic format would be selected, the file name is completely ignored and the extension of the selected graphic format would be appended to the entered file name and will be saved in this format.
- If the filename is empty, the saving process would be aborted.
- If the file exists, it would be best to ask whether it should be overwritten, if not call the Save As dialog again.
- It would only save the selected file extension entry in the Save As dialog and restore this setting the next time._

Please review and give remarks

Based on: https://stackoverflow.com/questions/9822177/is-there-a-way-to-automatically-add-extensions-to-a-file-using-qfiledialog-on-li

### CODE REVIEW

**Comment by borgmanJeremy** [comment]
>Solution is based on some discussion found in internet (it is know issue in QT), so code may look a little bit complicated, but i don't see a simpler way. If you have an idea to simplify a code - please add remars in review.

Can you link to the discussions?

**Comment by ZetaoYang** [comment]
Tested this PR from CI build, Win64,
image
When cancel "Save" window,
image

**Comment by borgmanJeremy** [comment]
@TomaszPluta Can you add a reference to the stack overflow link to the comments? It's required per the stack overflow license: https://stackoverflow.com/help/licensing

File: src/utils/screenshotsaver.cpp [hunk 1]
```diff
@@ -99,6 +99,55 @@ bool ScreenshotSaver::saveToFilesystem(const QPixmap& capture,
...
+{
+#if defined(Q_WS_WIN) || defined(Q_WS_MAC)
+    return QFileDialog::getSaveFileName(parent, title, directory, filter);
+#else
+    QFileDialog dialog(parent, title, directory, filter);
+    if (parent) {
+        dialog.setWindowModality(Qt::WindowModal);
+    }
+
+    QRegExp filter_regex(QLatin1String("(?:^\\*\\.(?!.*\\()|\\(\\*\\.)(\\w+)"));
```
Comment (borgmanJeremy) [inline src/utils/screenshotsaver.cpp:115]: Could you define this regex as a variable name, move to a function, or add a comment to explain what it does? I'm not a regex wizard :)

File: src/utils/screenshotsaver.cpp [hunk 2]
```diff
@@ -99,6 +99,55 @@ bool ScreenshotSaver::saveToFilesystem(const QPixmap& capture,
...
+    dialog.setAcceptMode(QFileDialog::AcceptSave);
+
+    dialog.selectNameFilter(ConfigHandler().getSaveAsFileExtension());
+    if (dialog.exec() == QDialog::Accepted) {
+
+        ConfigHandler().setSaveAsFileExtension(dialog.selectedNameFilter());
+        QString file_name = dialog.selectedFiles().first();
+        QFileInfo info(file_name);
+
+        if ((dialog.selectedNameFilter() == "By extension [default: *.png]")) {
```
Comment (borgmanJeremy) [inline src/utils/screenshotsaver.cpp:127]: "By extension [default: *.png]" should be defined as a variable so when it is re-used on line 174 there is no chance of it only changing in once place causing a future bug.

**Review by borgmanJeremy — CHANGES_REQUESTED** [review]

File: src/utils/screenshotsaver.cpp (hunk 1 above)
Comment (TomaszPluta) [reply src/utils/screenshotsaver.cpp]: I changed the regex to some QString operations (extracting extension from chosen filter). IMHO is easier to understand it now.

File: src/utils/screenshotsaver.cpp (hunk 2 above)
Comment (TomaszPluta) [reply src/utils/screenshotsaver.cpp]: Yeah, good point. I've moved it to const variables

**Review by borgmanJeremy — APPROVED** [review]

**Comment by borgmanJeremy** [comment]
Awesome, thanks for making the changes! Could you run clang-format and then I'll get it merged?

**Pull request merged**